### PR TITLE
Specify charset in `WWW-Authenticate` for Basic Auth

### PR DESCRIPTION
**Pull request by therepanic**
In this commit, we add support for the charset from [RFC-7617](https://datatracker.ietf.org/doc/html/rfc7617), which definitely solves the problem when the client does not know what charset we are parsing with.

Closes: gh-18755

### CODE REVIEW

File: web/src/main/java/org/springframework/security/web/authentication/www/BasicAuthenticationFilter.java [hunk 1]
```diff
@@ -353,7 +354,11 @@ public void setCredentialsCharset(String credentialsCharset) {
 		Assert.hasText(credentialsCharset, "credentialsCharset cannot be null or empty");
 		this.credentialsCharset = credentialsCharset;
 		if (this.authenticationConverter instanceof BasicAuthenticationConverter basicAuthenticationConverter) {
-			basicAuthenticationConverter.setCredentialsCharset(Charset.forName(credentialsCharset));
+			Charset charset = Charset.forName(credentialsCharset);
+			basicAuthenticationConverter.setCredentialsCharset(charset);
+			if (this.authenticationEntryPoint instanceof BasicAuthenticationEntryPoint basicAuthenticationEntryPoint) {
+				basicAuthenticationEntryPoint.setCharset(charset);
+			}
 		}
```
Comment (therepanic) [inline web/src/main/java/org/springframework/security/web/authentication/www/BasicAuthenticationFilter.java:357]: BTW, I think this is a cool solution for charset forwarding, just my 50 cents.

File: web/src/main/java/org/springframework/security/web/authentication/www/BasicAuthenticationEntryPoint.java [hunk 2]
```diff
@@ -53,7 +58,8 @@ public void afterPropertiesSet() {
 	@Override
 	public void commence(HttpServletRequest request, HttpServletResponse response,
 			AuthenticationException authException) throws IOException {
-		response.setHeader("WWW-Authenticate", "Basic realm=\"" + this.realmName + "\"");
+		String header = "Basic realm=\"" + this.realmName + "\", charset=\"" + this.charset.name() + "\"";
+		response.setHeader("WWW-Authenticate", header);
```
Comment (therepanic) [inline web/src/main/java/org/springframework/security/web/authentication/www/BasicAuthenticationEntryPoint.java]: [RFC-7617](https://datatracker.ietf.org/doc/html/rfc7617) states that charset is recommended but not mandatory.

This essentially leads us to different conclusions. Ideally, we should make it nullable to comply with this RFC. However, on the other hand, I see no reason to go with this convention. In what I believe is an elegant solution I've presented, unfortunately, we can't pass a nullable charset because `BasicAuthenticationFilter#setCredentialsCharset` doesn't accept a nullable value, and I certainly don't think we should change the contract for this reason. I'd like to hear other opinions on this matter, in case anyone needs it.

File: web/src/test/java/org/springframework/security/web/authentication/www/BasicAuthenticationEntryPointTests.java [hunk 3]
```diff
@@ -77,7 +78,18 @@ void commenceWhenResponseHasHeaderThenOverride() throws IOException {
 		ep.commence(request, response, new DisabledException("Disabled"));
 		List<String> headers = response.getHeaders("WWW-Authenticate");
 		assertThat(headers).hasSize(1);
-		assertThat(headers.get(0)).isEqualTo("Basic realm=\"hello\"");
+		assertThat(headers.get(0)).isEqualTo("Basic realm=\"hello\", charset=\"UTF-8\"");
```
Comment (therepanic) [inline web/src/test/java/org/springframework/security/web/authentication/www/BasicAuthenticationEntryPointTests.java:82]: I believe this isn't a breaking change. Clients will ignore the header parameters, and this is just a convention of our test.

**Comment by therepanic** [comment]
It turned out that some tests also required modifications and the addition of a charset. I wrote about this here: https://github.com/spring-projects/spring-security/pull/18760#discussion_r2823934023 and here: https://github.com/spring-projects/spring-security/pull/18760#discussion_r2823916764.

And over time, I'm starting to think that maybe the tests should still pass in this case? Then we'll have to create some kind of toggle switch, so we can enable charset additions. I'd like to know the best way to do this to fit the project.

I'm not sure what's best.

File: web/src/main/java/org/springframework/security/web/authentication/www/BasicAuthenticationFilter.java (hunk 1 above)
Comment (jzheaux) [reply web/src/main/java/org/springframework/security/web/authentication/www/BasicAuthenticationFilter.java:357]: Thanks for the creativity here. I think my hestitation is that this method is about setting the charset for the credentials. 

While I imagine it quite common that applications will be happy to have the charset for the credentials as well as the `WWW-Authenticate` header the same, I also imagine the surprise of learning that when I call `setCredentialsCharset`, it also sets charsets for non-credential-related attributes.

Instead, could we leave this out of the setter for now and rely on applications configuring the `BasicAuthenticationEntryPoint`? I think this is fine since I think the main reason applications will have to change it will be to set it to null to restore earlier behavior.

File: web/src/main/java/org/springframework/security/web/authentication/www/BasicAuthenticationEntryPoint.java [hunk 4]
```diff
@@ -40,11 +42,14 @@
  * authorized, causing it to prompt the user to login again.
  *
  * @author Ben Alex
+ * @author Andrey Litvitski
  */
 public class BasicAuthenticationEntryPoint implements AuthenticationEntryPoint, InitializingBean {
 
 	private @Nullable String realmName;
 
+	private Charset charset = StandardCharsets.UTF_8;
```
Comment (jzheaux) [inline web/src/main/java/org/springframework/security/web/authentication/www/BasicAuthenticationEntryPoint.java]: Since it's optional, let's mark it and its setter as `@Nullable`, check for null in `commence`, and add a test that confirms that it can be set to `null`.

File: web/src/main/java/org/springframework/security/web/authentication/www/BasicAuthenticationEntryPoint.java [hunk 5]
```diff
@@ -65,4 +71,12 @@ public void setRealmName(String realmName) {
 		this.realmName = realmName;
 	}
 
+	public void setCharset(Charset charset) {
+		this.charset = charset;
+	}
+
+	public Charset getCharset() {
```
Comment (jzheaux) [inline web/src/main/java/org/springframework/security/web/authentication/www/BasicAuthenticationEntryPoint.java]: Let's please leave the getter our for now

File: web/src/main/java/org/springframework/security/web/authentication/www/BasicAuthenticationEntryPoint.java [hunk 6]
```diff
@@ -65,4 +71,12 @@ public void setRealmName(String realmName) {
 		this.realmName = realmName;
 	}
 
+	public void setCharset(Charset charset) {
```
Comment (jzheaux) [inline web/src/main/java/org/springframework/security/web/authentication/www/BasicAuthenticationEntryPoint.java]: Please add a JavaDoc that indicates what the default value is. It should also explain that an application can set the value to `null` to remove the `charset` header attribute.

**Comment by therepanic** [comment]
Thanks for review. I followed your suggestions and pushed the change. I agree with your suggestions and followed them, pushing the change.

File: web/src/main/java/org/springframework/security/web/authentication/www/BasicAuthenticationEntryPoint.java [hunk 7]
```diff
@@ -65,4 +74,14 @@ public void setRealmName(String realmName) {
 		this.realmName = realmName;
 	}
 
+	/**
+	 * Sets the charset to include in the {@code WWW-Authenticate} response header. By
+	 * default, it is set to {@link StandardCharsets#UTF_8}. Set to {@code null} to omit
+	 * the charset attribute from the header.
+	 * @param charset the charset to use, or {@code null} to remove the charset attribute
+	 */
+	public void setCharset(@Nullable Charset charset) {
```
Comment (jzheaux) [inline web/src/main/java/org/springframework/security/web/authentication/www/BasicAuthenticationEntryPoint.java:85]: Since RFC 7617 only allows UTF-8, let's assert that `charset` is either `null` or UTF-8. We can always relax that requirement in the future.

File: web/src/main/java/org/springframework/security/web/authentication/www/BasicAuthenticationEntryPoint.java [hunk 8]
```diff
@@ -53,7 +58,11 @@ public void afterPropertiesSet() {
 	@Override
 	public void commence(HttpServletRequest request, HttpServletResponse response,
 			AuthenticationException authException) throws IOException {
-		response.setHeader("WWW-Authenticate", "Basic realm=\"" + this.realmName + "\"");
+		String header = "Basic realm=\"" + this.realmName;
```
Comment (jzheaux) [inline web/src/main/java/org/springframework/security/web/authentication/www/BasicAuthenticationEntryPoint.java]: This seems like a bug. If `this.charset` is `null`, then `header` will be missing a trailing quote.

File: web/src/test/java/org/springframework/security/web/authentication/www/BasicAuthenticationEntryPointTests.java [hunk 9]
```diff
@@ -62,7 +64,7 @@ public void testNormalOperation() throws Exception {
 		ep.commence(request, response, new DisabledException("These are the jokes kid"));
 		assertThat(response.getStatus()).isEqualTo(401);
 		assertThat(response.getErrorMessage()).isEqualTo(HttpStatus.UNAUTHORIZED.getReasonPhrase());
-		assertThat(response.getHeader("WWW-Authenticate")).isEqualTo("Basic realm=\"hello\"");
+		assertThat(response.getHeader("WWW-Authenticate")).isEqualTo("Basic realm=\"hello\", charset=\"UTF-8\"");
```
Comment (jzheaux) [inline web/src/test/java/org/springframework/security/web/authentication/www/BasicAuthenticationEntryPointTests.java:67]: Can you please add a test for `setCharset(null)`?

**Comment by jzheaux** [comment]
Thanks for the updates, @therepanic! I've also polished the documentation a bit, including mentioning this in `whats-new.adoc` so folks can be aware of the change. This will merge once the build passes.

**Review by jzheaux — APPROVED** [review]
